docs/architecture/adr-038-state-listening.md

```diff
@@ -269,162 +367,189 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 
 	...
 
-	// call the hooks with the BeginBlock messages
-	wg := new(sync.WaitGroup)
-	for _, streamingListener := range app.abciListeners {
-		streamingListener := streamingListener // https://go.dev/doc/faq#closures_and_goroutines
-		if streamingListener.HaltAppOnDeliveryError() {
-			// increment the wait group counter
-			wg.Add(1)
-			go func() {
-				// decrement the counter when the go routine completes
-				defer wg.Done()
-				if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
-					app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
-					app.halt()
-				}
-			}()
-		} else {
-			// fire and forget semantics
-			go func() {
-				if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
-					app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
-				}
-			}()
+	// call the streaming service hook with the BeginBlock messages
+	if app.abciListener != nil {
+		reqBz, err := req.Marshal()
```

Comment:
is this blocking? the previous design had it in a goroutine, here it looks like it could block continuation

Reply:
Correct! I need to fix it to maintain the previous design.

Reply:
This is blocking iff `abciListener` is not nil, that is the app is opted-in to streaming, correct?

Should all panics not roll up under `stopNodeOnStreamingErr` instead of just the `Listen*` result?